arborist/resource.go

```diff
@@ -0,0 +1,44 @@
+package arborist
+
+import ()
+
+// Representation of a resource (thing to which access is allowed). Resources
+// also have a hierarchical organization, where the children are meant to be
+// finer-grained levels of authorization. Resources must be uniquely identified
+// and the auth engine must maintain the collection of resource IDs.
+type Resource struct {
+	ID           string
+	subresources map[*Resource]struct{}
```

Comment:
why is this a map?

Reply:
oh I see the reasoning in subroles

Reply:
This is actually kind of clever using a map's keys as a set.